content/posts/2018-11-runtime-moonspeak.md

```diff
@@ -0,0 +1,131 @@
+
+But then, somewhere in the middle of localizing everything, we gave up.
+Turns out the perf hit of rendering a razor view for every string you want to render, ever, was just to big.
+We also needed to localize `.cs` files.
+
+[TODO IIRC there should be benchmark screenshots somewhere in hardcore][]
+
+Enter `roslyn` tooling, again, but with a twist!
+At that time, `csc.exe` was still the default C# compiler, and `aspnet_compiler.exe` was (and still is super slow).
+So we replaced _both_ of them, with the `roslyn`-based [StackExchange.Precompilation](https://github.com/StackExchange/StackExchange.Precompilation), and to top it of, we made it extensible so it could handle all of our localization needs (both extracting the strings, as well as precompiling localization in both `.cshtml` and `.cs` files).
+
+You can read more about this whole project [here](https://stackoverflow.blog/2015/07/23/announcing-stackexchange-precompilation/).
+
+# .NET Core
+
+We currently still run on ASPNET MVC 5, on the full framework. Our ultimate goal is to be able to run on .NET Core. AspNetCore is the first stepping stone towards that.
+
+## Present
+
+As we were evaluating ASPNET vNext (== 5+), we had always hoped to roll with the metaprogramming features that were available in the previews, either there, or what was worked on in [`roslyn` generators](TODO).
+But the time to move to .NET Core, and ASPNET DotNetCore is now, and unfortunately none of those features are around anymore / yet.
+Luckily, though, .NET tooling introduced some high-impact perf improvements that make the CLI tools much faster (e.g. `dotnet build-servers`).
+Additionally, AspNetCore `razor` view pre-compilation is fast!
+
+We had to future-proof our tooling, while still maintaining backwards compatibility with our old MVC stack.
+
+Our localization needs are still the same:
+
+### 1. No perf regressions
+Our code-gen wasn't perfect.
+It relied on the `roslyn` `SyntaxTree` only, we didn't use the semantic model much.
+A lot of the generated code was not optimal (e.g. we passed around all of the template properties as `object`s), but calling it was still way faster that having the entire `razor` pipeline on the stack.
+This gave us some wiggle room for writing an optimized runtime implementation.
+And we managed to make the runtime implementation as fast as the old (albeit not very optimized) compile-time code-gen-ed localization.
+Kudos to [@marcgravell](https://twitter.com/marcgravell), [@Nick_Craver](https://twitter.com/Nick_Craver) for helping out here. [BenchMarkDotNet](https://github.com/dotnet/BenchmarkDotNet) was also insanely helpful.
+
+### 1. Compile-time verification of string templates / parameters
+
+We've already written `roslyn` analyzers that did this.
+Since running the whole compilation with ASP.NET MVC 5 ~2 minutes on a beefy machine, was really a show stopper for developers, we wanted to give them instant-feedback inside the VS, which squiglies and all.
```

Comment:
```suggestion
Since running the whole compilation with ASP.NET MVC 5 took ~2 minutes on a beefy machine, was really a show stopper for developers, we wanted to give them instant-feedback inside VS, with squiglies and all.
```